plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StdLibConverter.kt

```diff
@@ -17,69 +16,192 @@ fun VariableEmbedding.sameSize(): ExpEmbedding =
 fun VariableEmbedding.increasedSize(amount: Int): ExpEmbedding =
     EqCmp(FieldAccess(this, ListSizeFieldEmbedding), Add(Old(FieldAccess(this, ListSizeFieldEmbedding)), IntLit(amount)))
 
-fun NamedFunctionSignature.stdLibPreConditions(): List<ExpEmbedding> =
-    NameMatcher.match(this.name) {
-        ifInCollectionsPkg {
-            ifFunctionName("get") {
-                val receiver = receiver!!
-                val indexArg = formalArgs[1]
-                return listOf(
-                    GeCmp(
-                        indexArg,
-                        IntLit(0),
-                        SourceRole.ListElementAccessCheck(SourceRole.ListElementAccessCheck.AccessCheckType.LESS_THAN_ZERO)
-                    ),
-                    GtCmp(
-                        FieldAccess(receiver, ListSizeFieldEmbedding),
-                        indexArg,
-                        SourceRole.ListElementAccessCheck(SourceRole.ListElementAccessCheck.AccessCheckType.GREATER_THAN_LIST_SIZE)
-                    ),
-                )
+sealed interface StdLibReceiverInterface {
+    fun match(function: NamedFunctionSignature): Boolean
+}
+
+sealed interface PresentInterface : StdLibReceiverInterface {
+    val interfaceName: String
+    override fun match(function: NamedFunctionSignature): Boolean =
+        function.receiverType?.isInheritorOfCollectionTypeNamed(interfaceName) == true
+}
+
+data object CollectionInterface : PresentInterface {
+    override val interfaceName = "Collection"
+}
+
+data object ListInterface : PresentInterface {
+    override val interfaceName = "List"
+}
+
+data object MutableListInterface : PresentInterface {
+    override val interfaceName = "MutableList"
+}
+
+data object NoInterface : StdLibReceiverInterface {
```

Comment:
I'm a bit confused, what is the goal of this?

Reply:
It was created for the `emptyList` function which is not a member function.

Reply:
But I think in the end we have too many problems with this __always writable__ field.
For example, we can't normally access `Map.entries` twice because we will have permission to `Set.size` field of the same `Set` twice and so on.

When we have some uniqueness system implemented, for we will probably be able reason about immutable lists since we have quite a lot of methods to create "correct" and unique (or at least without refs to mutable versions) lists (including `listOf`, `emptyList`, `toList`, `filter` and so on).

Reply:
Ah, I see.  If you need to, feel free to disable support for it for now.